pkl-core/src/main/java/org/pkl/core/runtime/VmListingOrMapping.java

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.pkl.core.runtime;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.frame.MaterializedFrame;
+import com.oracle.truffle.api.nodes.IndirectCallNode;
+import org.graalvm.collections.EconomicMap;
+import org.graalvm.collections.UnmodifiableEconomicMap;
+import org.pkl.core.PklBugException;
+import org.pkl.core.ast.member.ListingOrMappingTypeCheckNode;
+import org.pkl.core.ast.member.ObjectMember;
+import org.pkl.core.util.EconomicMaps;
+import org.pkl.core.util.Nullable;
+
+public abstract class VmListingOrMapping<SELF extends VmListingOrMapping<SELF>> extends VmObject {
+
+  /** The original that this listing/mapping is a surrogate of. It might have its own surrogatee. */
+  private final @Nullable SELF surrogatee;
```

Comment:
On the surrogate(e) naming; what about s/surrogate/typeCheckedThis and s/surrogatee/uncheckedThis?